main/src/main/scala/sbt/internal/InMemoryCacheStore.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * sbt
+ * Copyright 2011 - 2018, Lightbend, Inc.
+ * Copyright 2008 - 2010, Mark Harrah
+ * Licensed under Apache License 2.0 (see LICENSE)
+ */
+
+package sbt.internal
+
+import java.io.IOException
+import java.lang.Math.toIntExact
+import java.nio.file.attribute.BasicFileAttributes
+import java.nio.file.{ Files, Path }
+import java.util.concurrent.atomic.AtomicReference
+
+import com.github.benmanes.caffeine.cache.{ Cache, Caffeine, Weigher }
+import sbt.io.IO
+import sbt.util.{ CacheStore, CacheStoreFactory, DirectoryStoreFactory }
+import sjsonnew.{ IsoString, JsonReader, JsonWriter, SupportConverter }
+
+private[sbt] object InMemoryCacheStore {
+  private[this] class InMemoryCacheStore(maxSize: Long) extends AutoCloseable {
+    private[this] val weigher: Weigher[Path, (Any, Long, Int)] = { case (_, (_, _, size)) => size }
+    private[this] val files: Cache[Path, (Any, Long, Int)] = Caffeine
+      .newBuilder()
+      .maximumWeight(maxSize)
+      .weigher(weigher)
+      .build()
+    def get(path: Path): Option[(Any, Long)] = {
+      files.getIfPresent(path) match {
+        case null                     => None
+        case (value, lastModified, _) => Some((value, lastModified))
+      }
+    }
+    def put(path: Path, value: Any, lastModified: Long): Unit = {
+      try {
+        if (lastModified > 0) {
+          val attributes = Files.readAttributes(path, classOf[BasicFileAttributes])
+          files.put(path, (value, lastModified, toIntExact(attributes.size)))
+        }
+      } catch {
+        case _: IOException | _: ArithmeticException => files.invalidate(path)
+      }
+    }
+    def remove(path: Path): Unit = files.invalidate(path)
+
+    override def close(): Unit = {
+      files.invalidateAll()
+      files.cleanUp()
+    }
+  }
+
+  private[this] class CacheStoreImpl(path: Path, store: InMemoryCacheStore, cacheStore: CacheStore)
+      extends CacheStore {
+    override def delete(): Unit = cacheStore.delete()
+    override def read[T]()(implicit reader: JsonReader[T]): T = {
+      val lastModified = IO.getModifiedTimeOrZero(path.toFile)
+      store.get(path) match {
+        case Some((value: T, `lastModified`)) => value
+        case _                                => cacheStore.read[T]()
+      }
+    }
+    override def write[T](value: T)(implicit writer: JsonWriter[T]): Unit = {
+      val lastModified = IO.getModifiedTimeOrZero(path.toFile)
+      store.get(path) match {
+        case Some((v: T, `lastModified`)) if v == value => // nothing has changed
+        case _ =>
+          store.remove(path)
```

Comment:
You might want to use a `cache.asMap().compute` to avoid races. This way if multiple threads read a stale timestamp and try to reinsert, only one has to do the work. You could do something like,

```java
cache.asMap().compute(key, oldValue) -> {
  if old value is absent or stale
    // compute new value
    return newValue
  else
    return oldValue
})
```

Reply:
Thanks for the suggestion! I was looking at the javadocs for ConcurrentMap and it looks like I'd also need to add some synchronization to prevent multiple threads from concurrently writing the same file. Based on the [javadocs](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentMap.html#compute-K-java.util.function.BiFunction-), it looks like this could actually cause more work to be done:
> The default implementation is equivalent to performing the following steps for this map, then returning the current value or null if absent:
>
>
>```java
> V oldValue = map.get(key);
> V newValue = remappingFunction.apply(key, oldValue);
> if (oldValue != null ) {
>    if (newValue != null)
>       map.replace(key, oldValue, newValue);
>    else
>       map.remove(key, oldValue);
> } else {
>    if (newValue != null)
>       map.putIfAbsent(key, newValue);
>    else
>       return null;
 >}
>```
> 
>The default implementation may retry these steps when multiple threads attempt updates including potentially calling the remapping function multiple times.

The way this cache is used, I don't think that multiple threads should really ever be trying to modify the same file. The CacheStore provides a read/write api for a specific file corresponding to a task. For example, the previous value of a task. Since a task is only run once per command evaluation and sbt doesn't run commands concurrently, these files should only be written to at most once per command. It is a low throughput cache. I'll maybe add a comment about concurrency though on the chance that this becomes an issue.

Reply:
The default implementation is not atomic due to backwards compatibility, etc. However `ConcurrentHashMap` and `Caffeine' do implement it atomically so it is safe to rely on. You just have to be mindful when using other `ConcurrentMap` types, like `ConcurrentSkipListMap` or external ones like `NonBlockingHashMap` which do not support atomic computations.

If it's not an issue due to how you are using it, that's even better.

Reply:
>Attempts to compute a mapping for the specified key and its current mapped value (or null if there is no current mapping). The entire method invocation is performed atomically. Some attempted update operations on this map by other threads may be blocked while computation is in progress, so the computation should be short and simple, and must not attempt to update any other mappings of this Map.

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-

Reply:
Aah, cool. Thanks for the clarification.